Regularize mask handling for attn_mask and key_padding_mask

mikekgfb opened the pull request:
Summary:
Regularize mask handling for attn_mask and key_padding_mask
* Update documentation to remove reference to byte masks (which were deprecated long ago)
* Introduce check and warn about deprecation if attn_mask and key_padding_mask types mismatch
* Convert all masks to float before combining
* Combine by adding

Test Plan: sandcastle & github CI

Differential Revision: D42653215

ngimel commented on torch/nn/modules/activation.py:
```diff
@@ -1037,7 +1037,7 @@ def forward(
         key_padding_mask: If specified, a mask of shape :math:`(N, S)` indicating which elements within ``key``
             to ignore for the purpose of attention (i.e. treat as "padding"). For unbatched `query`, shape should be :math:`(S)`.
-            Binary and byte masks are supported.
+            Binary and floati masks are supported.
```
```suggestion
            Binary and float masks are supported.
```

drisspg commented on torch/nn/functional.py:
```diff
@@ -5058,6 +5054,18 @@ def multi_head_attention_forward(
                 "only bool and floating types of key_padding_mask are supported")
+        if attn_mask is not None:
+            if _kpm_dtype != attn_mask.dtype:
+                warnings.warn("Support for mismatched attn_mask and kay_padding_mask in nn.MultiheadAttention is deprecated. Use same type for both instead.")
```
nit: key

ngimel commented:
I'm a bit worried that we need to put in key_padding_mask conversion in so many places (and btw I think it's also needed in nn/modules/transformer.py in TransformerEncoder module), could we create a helper function that would be called from everywhere if we don't want a more serious refactor? 
There is a small perf impact from conversion but I think that's fine. 
What about tests?

mikekgfb commented:
> I'm a bit worried that we need to put in key_padding_mask conversion in so many places (and btw I think it's also needed in nn/modules/transformer.py in TransformerEncoder module), could we create a helper function that would be called from everywhere if we don't want a more serious refactor?

We had a more serious refactor, but it collided with landing sdpa() which was a priority.  Totally agree it is desirable.  Will be good to find somebody to do that.

> could we create a helper function that would be called from everywhere if we don't want a more serious refactor?

Can we do this as a follow-up task?  It's getting unwieldy and I'd like to land a solution.  (I'd desist from fixing transformer encoder and land that concurrently with the helper function update.)

> What about tests?

Transformers and Attention are already tested, right?  So, simplifying the implementation shouldn't require new tests?
(I'll create a task to track creating more tests, because clearly we missed some combinations.)

ngimel commented:
But existing tests didn't uncover this bug, so we need new ones?

mikekgfb commented:
Refactored the testing and implemented canonical mask conversion across F.MHA, nn.MHA, and nn.TransformerEncoder.  Opens the path for a bigger refactor to validate and combine masks as a common building block in a future update.

> But existing tests didn't uncover this bug, so we need new ones?

I added a suitable test.

cpuhrsch commented on torch/nn/functional.py:
```diff
@@ -4910,6 +4910,41 @@ def _mha_shape_check(query: Tensor, key: Tensor, value: Tensor,
 
     return is_batched
 
+def canonical_mask(
```
Can we add an underscore since these are not public functions?

That is `_canonical_mask` instead of `canonical_mask`

mikekgfb replied:
Done

cpuhrsch commented on torch/nn/functional.py:
```diff
@@ -4910,6 +4910,41 @@ def _mha_shape_check(query: Tensor, key: Tensor, value: Tensor,
+            )
+    return mask
+
+def none_or_dtype(input: Optional[Tensor]) -> Optional[DType]:
```
Same here

mikekgfb replied:
Done

mikekgfb replied:
Approval please?

drisspg commented on test/test_transformers.py:
```diff
@@ -1454,6 +1454,25 @@ def test_is_causal_gpu(self):
+        mha = nn.MultiheadAttention(E, H)
+        mha.in_proj_weight = Parameter(torch.ones((E * 3, E)))
+        mha.out_proj.weight = Parameter(torch.ones((E, E)))
+        expected = torch.ones(size=(S, L, E)) * 16
```
nit: is this used?
Are is this test just to ensure that kpm and am can't both be set? if so I think the test name should somewhat reflect that

drisspg commented on torch/nn/modules/transformer.py:
```diff
@@ -720,6 +723,7 @@ def _ff_block(self, x: Tensor) -> Tensor:
 
 
 def _get_clones(module, N):
+    # FIXME: copy.deepcopy() is not defined on nn.module
```
Is there an issue for this?

Commit:
Regularize mask handling for attn_mask and key_padding_mask (#92733)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/92733

Regularize mask handling for attn_mask and key_padding_mask
* Update documentation to remove reference to byte masks (which were deprecated long ago)
* Introduce check and warn about deprecation if attn_mask and key_padding_mask types mismatch
* Convert all masks to float before combining
* Combine by adding

Test Plan: sandcastle & github CI

Reviewed By: ngimel

Differential Revision: D42653215

fbshipit-source-id: 5575652967ff2660c16e387c5de11d2af4ae4a19

mikekgfb commented:
@pytorchbot merge

mikekgfb commented:
@pytorchbot merge
